### PR TITLE
Token system: Decimal / limit attributes implementation

### PR DESCRIPTION
**Pull request by bvbfan**
Mostly a WIP it should correct handle different token precision width as well as maximum token value to be user defined. I did not decide which is better approach, querying storage for token info, as this approach or adding token info to pool pair and (not serializable) but updated in `GetPoolPair` when needed. Two different tokens are adjusted by higher precision one, pool pair token as well, liquidity is downscale to lower precision token when needed. Unit tests pass, functional tests are still not performed. Suggestions are welcome, better design, better handling token precision / limit.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2594,7 +2611,7 @@ void CWallet::AvailableCoins(interfaces::Chain::Lock& locked_chain, std::vector<
             vCoins.push_back(COutput(&wtx, i, nDepth, spendable, solvable, safeTx, (coinControl && coinControl->fAllowWatchOnly)));
 
             // Checks the sum amount of all UTXO's.
-            if (nMinimumSumAmount != MAX_MONEY) {
+            if (nMinimumSumAmount != INT64_MAX) {
```
Comment (ShengguangXiao) [inline src/wallet/wallet.cpp:2614]: This change will break the DFI consensus.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3026,7 +3043,7 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
             CCoinControl ccSingleToken(coin_control);
             ccSingleToken.m_tokenFilter = { tokenId };
 
-            AvailableCoins(*locked_chain, vAvailableCoins, true, &ccSingleToken, 1, MAX_MONEY, MAX_MONEY, 0);
+            AvailableCoins(*locked_chain, vAvailableCoins, true, &ccSingleToken, 1, INT64_MAX, INT64_MAX, 0);
```
Comment (ShengguangXiao) [inline src/wallet/wallet.cpp:3046]: This change will break the DFI UTXO consensus.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3083,7 +3100,7 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
         LOCK(cs_wallet);
         {
             std::vector<COutput> vAvailableCoins;
-            AvailableCoins(*locked_chain, vAvailableCoins, true, &coin_control, 1, MAX_MONEY, MAX_MONEY, 0);
+            AvailableCoins(*locked_chain, vAvailableCoins, true, &coin_control, 1, INT64_MAX, INT64_MAX, 0);
```
Comment (ShengguangXiao) [inline src/wallet/wallet.cpp:3103]: This change will break the DFI UTXO consensus.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -1555,10 +1555,14 @@ TAmounts CWallet::GetDebit(const CTransaction& tx, const isminefilter& filter) c
     for (const CTxIn& txin : tx.vin)
     {
         CTokenAmount const amount = GetDebit(txin, filter);
-        nDebit[amount.nTokenId] += amount.nValue;
-        if (!MoneyRange(nDebit.at(amount.nTokenId)))
+        auto& val = nDebit[amount.nTokenId];
+        auto res = SafeAdd(val, amount.nValue);
+        if (!res)
```
Comment (ShengguangXiao) [inline src/wallet/wallet.cpp:1560]: Here should use `if (!res.ok)`?

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -1589,10 +1593,14 @@ TAmounts CWallet::GetCredit(const CTransaction& tx, const isminefilter& filter)
     for (const CTxOut& txout : tx.vout)
     {
         CTokenAmount const amount = GetCredit(txout, filter);
-        nCredit[amount.nTokenId] += amount.nValue;
-        if (!MoneyRange(nCredit.at(amount.nTokenId)))
+        auto& val = nCredit[amount.nTokenId];
+        auto res = SafeAdd(val, amount.nValue);
+        if (!res)
```
Comment (ShengguangXiao) [inline src/wallet/wallet.cpp:1598]: Here should use if (!res.ok)?

File: src/masternodes/govvariables/lp_splits.cpp [hunk 6]
```diff
@@ -9,43 +9,56 @@
...
+        uint8_t decimal = token ? token->decimal : 8;
+        int64_t coin = std::pow(10, decimal);
+        if (kv.second < 0 || kv.second > coin)
             return Res::Err("wrong percentage for pool with id=%s, value = %s", kv.first.ToString(), std::to_string(kv.second));
 
-        total += kv.second;
+        coins[kv.first] = coin;
+        totals[kv.first] += kv.second;
+    }
+    for (auto& total : totals) {
```
Comment (ShengguangXiao) [inline src/masternodes/govvariables/lp_splits.cpp:58]: add `const` before auto.

File: src/masternodes/govvariables/lp_splits.cpp [hunk 7]
```diff
@@ -9,43 +9,56 @@
...
+    TAmounts totals;
+    std::map<DCT_ID, int64_t> coins;
     for (auto const & kv : splits) {
         auto pool = mnview.GetPoolPair(kv.first);
 
         if (!pool)
             return Res::Err("pool with id=%s not found", kv.first.ToString());
 
-        if (kv.second < 0 || kv.second > COIN)
+        auto token = mnview.GetToken(kv.first);
```
Comment (ShengguangXiao) [inline src/masternodes/govvariables/lp_splits.cpp:49]: add const

File: src/masternodes/govvariables/lp_splits.cpp [hunk 8]
```diff
@@ -9,43 +9,56 @@
...
+        uint8_t decimal = token ? token->decimal : 8;
+        int64_t coin = std::pow(10, decimal);
```
Comment (ShengguangXiao) [inline src/masternodes/govvariables/lp_splits.cpp:51]: Add const for these three values

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -2902,8 +2902,8 @@ static UniValue listunspent(const JSONRPCRequest& request)
     }
 
     CAmount nMinimumAmount = 0;
-    CAmount nMaximumAmount = MAX_MONEY;
-    CAmount nMinimumSumAmount = MAX_MONEY;
+    CAmount nMaximumAmount = INT64_MAX;
+    CAmount nMinimumSumAmount = INT64_MAX;
```
Comment (ShengguangXiao) [inline src/wallet/rpcwallet.cpp:2906]: This may break UTXO consensus

**Comment by monstrobishi** [comment]
@ShengguangXiao yes this is intended to break consensus it will be for an upcoming hardfork BUT the base should not be master it should be v1.4.x thanks for pointing that out

@bvbfan can you change the merge base to v1.4.x please

File: src/masternodes/mn_checks.cpp [hunk 10]
```diff
@@ -310,6 +310,17 @@ Res ApplyCreateTokenTx(CCustomCSView & mnview, CCoinsViewCache const & coins, CT
         }
     }
 
+    if (token.decimal > 18) {
+        return Res::Err("%s : %s", base, "Decimal is too deep");
+    }
+
+    CAmount coin = std::pow(10, token.decimal);
+    if (token.limit == 0) {
+        token.limit = 1200000000 * coin; // default token limit
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:319]: Should we use `MAX_MONEY` instead of `1200000000`?

File: src/Makefile.test.include [hunk 11]
```diff
@@ -185,7 +185,7 @@ DEFI_TEST_SUITE += \
 endif
 
 test_test_defi_SOURCES = $(DEFI_TEST_SUITE) $(DEFI_TESTS) $(JSON_TEST_FILES) $(RAW_TEST_FILES)
-test_test_defi_CPPFLAGS = $(AM_CPPFLAGS) $(DEFI_INCLUDES) $(TESTDEFS) $(EVENT_CFLAGS)
+test_test_defi_CPPFLAGS = $(AM_CPPFLAGS) $(DEFI_INCLUDES) $(TESTDEFS) $(EVENT_CFLAGS) -DTEST_BUILD
```
Comment (ShengguangXiao) [inline src/Makefile.test.include:188]: Should we push this `TEST_BUILD` flag into code base? Maybe we just use it personally.

Comment (bvbfan) [reply src/Makefile.test.include:188]: I'll remove it.

File: src/masternodes/mn_checks.cpp (hunk 10 above)
Comment (bvbfan) [reply src/masternodes/mn_checks.cpp:319]: It's correct `MAX_MONEY / COIN * coin` but it can overflow int64 it'll be handled better in next push.

**Comment by bvbfan** [comment]
- [ ] All unit and functional tests 

- [ ] New unit and functional test

**Comment by monstrobishi** [comment]
Conflicts need to be resolved

**Comment by bvbfan** [comment]
Now it should be ready for review.

File: src/consensus/tx_verify.cpp [hunk 12]
```diff
@@ -182,18 +182,30 @@ bool Consensus::CheckTxInputs(const CTransaction& tx, CValidationState& state, c
...
+        val = res;
         /// @todo tokens: later match the range with TotalSupply
-
         if (prevout.n == 1 && !mnview->CanSpend(prevout.hash, nSpendHeight)) {
             return state.Invalid(ValidationInvalidReason::CONSENSUS, false, REJECT_INVALID, "bad-txns-collateral-locked",
                 strprintf("tried to spend locked collateral for %s", prevout.hash.ToString())); /// @todo may be somehow place the height of unlocking?
         }
     }
 
+    for (auto& value : nValuesIn) {
```
Comment (ShengguangXiao) [inline src/consensus/tx_verify.cpp:197]: Add const

File: src/consensus/validation.h [hunk 13]
```diff
@@ -159,4 +159,13 @@ static inline int64_t GetTransactionInputWeight(const CTxIn& txin)
     return ::GetSerializeSize(txin, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS) * (WITNESS_SCALE_FACTOR - 1) + ::GetSerializeSize(txin, PROTOCOL_VERSION) + ::GetSerializeSize(txin.scriptWitness.stack, PROTOCOL_VERSION);
 }
 
+template <typename Callback>
+void ForEachTokenOutput(const CTransaction& tx, Callback callback)
+{
+    std::set<DCT_ID> tokens;
+    for (const auto& txout : tx.vout)
+        if (tokens.insert(txout.nTokenId).second)
+            callback(txout.nTokenId);
```
Comment (ShengguangXiao) [inline src/consensus/validation.h:168]: Can add braces here? In case future add other logic inside.

File: src/defi-tx.cpp [hunk 14]
```diff
@@ -614,8 +614,25 @@ static void MutateTxSign(CMutableTransaction& tx, const std::string& flagStr)
...
+                            const auto str = amount.getValStr();
+                            const auto pos = str.find('.');
+                            if (pos == str.npos)
+                                throw std::runtime_error("decimal point not present");
+                            decimal = str.size() - pos - 1;
+                        }
+                        if (decimal > 18)
+                            throw std::runtime_error("decimal is unrealistic");
+                    }
+                    newcoin.out.nValue = AmountFromValue(amount, decimal);
```
Comment (ShengguangXiao) [inline src/defi-tx.cpp:635]: Can put this section of code into a function?

File: src/masternodes/mn_checks.cpp [hunk 15]
```diff
@@ -310,6 +320,23 @@ Res ApplyCreateTokenTx(CCustomCSView & mnview, CCoinsViewCache const & coins, CT
         }
     }
 
+    if (token.decimal > 18) {
+        return Res::Err("%s : %s", base, "Decimal is too deep");
+    }
+
+    CAmount coin = std::pow(10, token.decimal);
+    if (token.limit == 0) {
+        auto limit = MAX_MONEY / COIN;
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:329]: Here should be auto limit = MAX_MONEY / coin? coin and COIN very easy to get confused, i suggest to rename one of them.

File: src/masternodes/mn_checks.cpp [hunk 16]
```diff
@@ -310,6 +320,23 @@ Res ApplyCreateTokenTx(CCustomCSView & mnview, CCoinsViewCache const & coins, CT
         }
     }
 
+    if (token.decimal > 18) {
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:323]: Define 18 as a const variable?

File: src/masternodes/mn_checks.cpp [hunk 17]
```diff
@@ -310,6 +320,23 @@ Res ApplyCreateTokenTx(CCustomCSView & mnview, CCoinsViewCache const & coins, CT
...
+        return Res::Err("%s : %s", base, "Decimal is too deep");
+    }
+
+    CAmount coin = std::pow(10, token.decimal);
+    if (token.limit == 0) {
+        auto limit = MAX_MONEY / COIN;
+        const auto cbits = bits(coin);
+        const auto lbits = bits(limit);
+        if (lbits + cbits > 63) { // overflow
```
Comment (ShengguangXiao) [inline src/masternodes/mn_checks.cpp:332]: Define 63 as a const variable?

File: src/masternodes/mn_rpc.cpp [hunk 18]
```diff
@@ -128,31 +129,35 @@ std::vector<CTxIn> GetInputs(UniValue const& inputs) {
...
+
+    pwallet->BlockUntilSyncedToCurrentChain();
+    auto locked_chain = pwallet->chain().lock();
+    LOCK(pwallet->cs_wallet);
+
+    std::vector<COutput> vecOutputs;
+    pwallet->AvailableCoins(*locked_chain, vecOutputs, true, &cctl, 1, INT64_MAX, INT64_MAX, 1);
+    if (vecOutputs.empty())
+        return false;
+
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:149]: Remove extra blank line

File: src/masternodes/mn_rpc.cpp [hunk 19]
```diff
@@ -547,9 +576,9 @@ UniValue createtoken(const JSONRPCRequest& request) {
                             {"isDAT", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED,
                              "Token's 'isDAT' property (bool, optional), default is 'False'"},
                             {"decimal", RPCArg::Type::NUM, RPCArg::Optional::OMITTED,
-                             "Token's decimal places (optional, fixed to 8 for now, unchecked)"},
+                             "Token's decimal places (optional, default to 8)"},
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:579]: default is 8

File: src/masternodes/mn_rpc.cpp [hunk 20]
```diff
@@ -613,11 +638,40 @@ UniValue createtoken(const JSONRPCRequest& request) {
...
+    }
+    if (!metaObj["mintable"].isNull()) {
+        token.flags = metaObj["mintable"].getBool() ?
+                        token.flags | (uint8_t)CToken::TokenFlags::Mintable :
+                        token.flags & ~(uint8_t)CToken::TokenFlags::Mintable;
+    }
+
+    if (!metaObj["decimal"].isNull()) {
+        auto decimal = metaObj["decimal"].get_int();
+        if (decimal > 18)
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:660]: Same as above, use const variable

File: src/masternodes/mn_rpc.cpp [hunk 21]
```diff
@@ -613,11 +638,40 @@ UniValue createtoken(const JSONRPCRequest& request) {
...
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "decimal is too deep");
+        token.decimal = uint8_t(decimal);
+    }
+
+    if (!metaObj["limit"].isNull()) {
+        auto limit = metaObj["limit"].get_int64();
+        if (limit < 1000) // pick some min value
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "limit should be greater value");
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:668]: The error message may need to refine

File: src/masternodes/mn_rpc.cpp [hunk 22]
```diff
@@ -613,11 +638,40 @@ UniValue createtoken(const JSONRPCRequest& request) {
...
+        auto diff = std::numeric_limits<int64_t>::max() / limit;
+        auto precision = std::pow(10, token.decimal);
+        if (precision > diff)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "limit and decimal are out of sync");
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:672]: The error message may need to refine

File: src/masternodes/mn_rpc.cpp [hunk 23]
```diff
@@ -911,7 +916,7 @@ UniValue tokenToJSON(DCT_ID const& id, CTokenImplementation const& token, bool v
         tokenObj.pushKV("isLPS", token.IsPoolShare());
         tokenObj.pushKV("finalized", token.IsFinalized());
 
-        tokenObj.pushKV("minted", ValueFromAmount(token.minted));
+        tokenObj.pushKV("minted", ValueFromAmount(token.minted, token.decimal));
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:919]: Why minted become decimal?

File: src/masternodes/mn_rpc.cpp [hunk 24]
```diff
@@ -1504,28 +1482,39 @@ UniValue poolToJSON(DCT_ID const& id, CPoolPair const& pool, CToken const& token
     poolObj.pushKV("idTokenB", pool.idTokenB.ToString());
 
     if (verbose) {
-        poolObj.pushKV("reserveA", ValueFromAmount(pool.reserveA));
-        poolObj.pushKV("reserveB", ValueFromAmount(pool.reserveB));
+        poolObj.pushKV("reserveA", ValueFromAmount(pool.reserveA, pool.tokenA.decimal));
+        poolObj.pushKV("reserveB", ValueFromAmount(pool.reserveB, pool.tokenB.decimal));
```
Comment (ShengguangXiao) [inline src/masternodes/mn_rpc.cpp:1486]: Why reserve becomes decimal?

File: src/masternodes/mn_checks.cpp (hunk 15 above)
Comment (bvbfan) [reply src/masternodes/mn_checks.cpp:329]: No, here is to determine what's the base value without precision.

File: src/masternodes/mn_rpc.cpp (hunk 23 above)
Comment (bvbfan) [reply src/masternodes/mn_rpc.cpp:919]: Correct. Mint in DFI only

File: src/masternodes/mn_rpc.cpp (hunk 24 above)
Comment (bvbfan) [reply src/masternodes/mn_rpc.cpp:1486]: Reserves are based on two token ids in pool pair.

Comment (ShengguangXiao) [reply src/masternodes/mn_rpc.cpp:1485]: Yes, it is correct, I only saw the second parameter of the input.

**Comment by ShengguangXiao** [comment]
@bvbfan Can you avoid force push after the review started? Now I cannot find what have been changed after comment.

**Comment by bvbfan** [comment]
Sure, the changes are requested ones, amount.h defines constants, defi-tx function to handle amount and precision. I've not prefer history to became garbage with pointless commits, so it is better to squash after review is done?

**Comment by ShengguangXiao** [comment]
Yes, we can squash when the PR is ready for merge

**Comment by uzyn** [comment]
@bvbfan @ShengguangXiao Can you confirm the cap for 18 decimal places (common for Ethereum)?

If it's just 1 (int64 size - 1e18), then there's no point on implementing this until we can support bigint. 

Recent encounter

```
# MOCK is 8 decimals

$ defi-cli minttokens '"10000000000@MOCK"'
error code: -3
error message:
: Invalid amount
$ defi-cli minttokens '"1000000000@MOCK"'
GOOD
```

**Comment by bvbfan** [comment]
`If it's just 1 (int64 size - 1e18)`
That's it, right.

**Comment by uzyn** [comment]
> `If it's just 1 (int64 size - 1e18)`
> That's it, right.

ok then it makes no sense for us to support variable decimals. If we do, I would recommend a max of 8 decimal places.

My opinion is that, take out the whole decimals implementation. Force it at 8.

**Comment by bvbfan** [comment]
> My opinion is that, take out the whole decimals implementation. Force it at 8.

It's forced to 8 till now (it was hardcoded in many places) this patch aims to handle better different decimal points and limits (which was hardcoded as well). So we can make a range of decimal point values say (2-10) and limit (?-?)

**Comment by monstrobishi** [comment]
@uzyn I think we close this, we decided not to go down this route right

**Comment by bvbfan** [comment]
I'll not delete the branch, so can return to it next time, it's well tested by me.

**Pull request closed without merging**